src/main/java/org/junitpioneer/jupiter/NotWorkingExtension.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2016-2021 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import static org.junit.jupiter.api.Assertions.fail;
+
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.Extension;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
+import org.junit.jupiter.api.extension.ExtensionContext.Store;
+import org.junit.jupiter.api.extension.LifecycleMethodExecutionExceptionHandler;
+import org.junit.jupiter.api.extension.TestExecutionExceptionHandler;
+import org.opentest4j.TestAbortedException;
+
+class NotWorkingExtension implements Extension, TestExecutionExceptionHandler, LifecycleMethodExecutionExceptionHandler,
+		BeforeEachCallback, AfterEachCallback {
```

Comment:
I was surprised by how many extension points are involved here and checked [the existing extension points](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/extension/Extension.html), looking for a simpler approach. I think I found it with `InvocationInterceptor`, where all you need to do is proceed with the invocation and then invert the result, i.e. fail on a successful execution and abort on failure. Here's a proof of concept:

```java
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.InvocationInterceptor;
import org.junit.jupiter.api.extension.ReflectiveInvocationContext;
import org.opentest4j.AssertionFailedError;
import org.opentest4j.TestAbortedException;

import java.lang.reflect.Method;

import static org.junit.jupiter.api.Assertions.fail;

public class LabTests {

	@Test
	@ExtendWith(InvertResultExtension.class)
	void test() {
		System.out.println("Hi!");
		fail();
	}

	private static class InvertResultExtension implements InvocationInterceptor {

		@Override
		public void interceptTestMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
			invert(invocation);
		}

		@Override
		public <T> T interceptTestFactoryMethod(Invocation<T> invocation, ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
			return invert(invocation);
		}

		private <T> T invert(Invocation<T> invocation) {
			try {
				invocation.proceed();
			} catch (Throwable ex) {
				throw new TestAbortedException("Test failed as expected.", ex);
			}
			throw new AssertionFailedError("Test unexpectedly passed.");
		}

	}

}
```

Reply:
That seems to work for nearly all cases except for exceptions thrown in an `@AfterEach` method, which would then probably require similarly complex logic as it is currently implemented.

Should I switch to `InvocationInterceptor` nonetheless?